README

   $Header: /var/cvs/yarrow/index.php,v 1.8 2002/12/08 16:14:39 marnanel
   Exp $
   
Yarrow: the manual

Contents

    1. [1]Introduction
    2. [2]Getting started as a user
         1. [3]Basics
         2. [4]GROGGS
         3. [5]A glossary
         4. [6]Finding out more
    3. [7]Getting started as an administrator
         1. [8]Installing yarrow
         2. [9]Installing your own RGTP servers
    4. [10]The future
    5. [11]Credits
       
1. Introduction

   This is the manual for yarrow. Yarrow lets you read RGTP bulletin
   boards, such as [12]GROGGS at Cambridge University, using a web
   browser.
   
   Some other pages worth looking at: [13]screen shots, [14]contact
   information, and [15]links to other RGTP sites.
   
2. Getting started as a user

2.1. Basics

   Yarrow is most commonly used with GROGGS, so I'll use that as an
   example here. The procedure is fairly similar for any RGTP board,
   though. You'll also need a copy of yarrow to play with; if you don't
   have one locally, you can use the example one at
   [16]http://rgtp.thurman.org.uk/gossip. There's also [17]a copy running
   inside the University Data Network, but that may not always be running
   the most recent version of the program. If you want to set up a copy
   on your own computer, skip ahead to the section "Getting started as an
   administrator".
   
   Firstly, it's useful to set up a yarrow account, so that all state
   (such as whether or not you've read a particular item) can be held on
   the yarrow server. It's not necessary to set one up; if you go to
   [18]the GROGGS index without having logged in, for example, you can
   simply click the "visitor" button to read the whole of GROGGS. (You
   might like to try doing that now for a bit.) However, you'll need an
   account if you want to post, so let's set one up: click "log in" on
   the right-hand menu, and then "get a new account". Enter your email
   address, which yarrow uses to identify you, and press OK.
   
   While you're waiting for the confirmation email from the server, let's
   discuss how RGTP works. An RGTP board consists of several discussion
   threads, called items. Any user with posting access can create new
   items, or add replies to the end of an item. Items have a fixed
   maximum size, so when one gets too long, you have to start another,
   with a name more pertinent to whatever the discussion's about now
   (this helps minimise the distance between subject lines and discussion
   topics that's so common on Usenet). The system tracks when a new item
   is a continuation of an old one, so it's possible to follow a long
   discussion through-- this is known as a thread.
   
   You should have your email from the server by now. Click the "change
   your password" link, and fill in the address you used, the password
   you were given, and two copies of a new password. Once you've clicked
   OK, you'll be logged into yarrow.
   
2.2. GROGGS

   Click the "gossip" link in this page (or just go back to [19]the main
   yarrow page) and you'll be presented with a list of servers. We want
   to read GROGGS, so click on the "groggs" link in the left-hand column.
   
   You're now looking at [20]the index of items on GROGGS-- well, the
   most recent of them: you can page back all the way to 1995, when
   GROGGS took its current form. Each item shows the date it was most
   recently added to (with the newer ones towards the top), the number
   ("#") of replies to that item, the name of the user who most recently
   added to it, and the subject. The items in bold, which will be all of
   them at present, are items which contain replies you haven't read;
   those in italics are full items, which will have been continued
   further up the index.
   
   Click any of the subject lines to read that item and all the replies.
   Each reply is preceded by a block giving the user's email address, and
   the date of posting, on the right. On the left is the user's grogname:
   a fanciful, possibly descriptive, piece of text which is associated
   with that particular user. If you were logged in as a regular user, at
   the bottom of the page there'd be [21]a box to add your own reply.
   
2.3. A glossary

   Spend a little time exploring GROGGS now. Here's some information to
   help you explore it.
   
   hup
          Most RGTP clients list the index with the items that have most
          recently been added to towards the top. Most groggers only read
          the top few items in the index. So if you feel an item deserves
          more discussion, you can append a reply saying "hup", which
          will bring it to the top of the index. You can theoretically
          hup any item (unless it's been continued), but doing so with
          items more than a few years old is considered odd.
          
   grogvote
          an annual contest to nominate groggers in various categories.
          (The categories are also nominated by the users, but "Star of
          GROGGS" and "Sage of GROGGS" are traditional.) The results are
          read out every year at the [22]GROGGS garden party, which
          happens during Cambridge May Week.
          
   [23]ucam.chat
          a newsgroup internal to Cambridge University. In recent years,
          it has taken a lot of the traffic (and traditions) which once
          were part of GROGGS.
          
   grogger; grogging
          a user of GROGGS; to use GROGGS
          
   GREED
          An important RGTP-reading program; it runs under Emacs. See
          [24]the GREED home page for more information.
          
   Infinity Questions (IQ)
          A very long-running game on GROGGS. One user picks a subject
          and the others have to ask "yes/no" questions to work out what
          it is. Rounds can sometimes last months. New players should be
          warned that the definition of "subject" is rather looser than
          in many similar games: the subjects of previous rounds have
          included: the autonomic nervous system; honesty; [25]PuTTY;
          cork (the substance); Cambridge, Massachusetts; the hosts and
          routers of the global Internet; Leo Blair; the London Eye; a
          backup tape for chiark.greenend.org.uk; and the White House's
          red encrypted telephone to the Kremlin.
          
   ITWSBT
          I think we should be told
          
   YA(x)AICMFP
          You are (x) and I claim my five pounds
          
   TIOSSUOTW(x)TIWPAO (and variants)
          [26]This is obviously some strange usage of the word (x) that I
          wasn't previously aware of
          
   SWITCH NOBOYDIE / SNB / DoeS aNyBody...
          [27]Chris Boyd (cb113) was a user of GROGGS who would always
          answer questions of the form "Does anyone know the way to
          Arbury?" with "Yes". To work around this, people began writing
          SWITCH NOBOYDIE before such questions, which soon became SNB.
          It's also often mixed into the question itself: "DoeS aNyBody
          know the way to Arbury?"
          
   spqr1
          any Cambridge person (since this is the general form of a
          Cambridge user id)
          
2.4. Finding out more

   If, now you've looked around, you'd like to participate in GROGGS, all
   you need to do is register for a GROGGS account. The link to do this
   is given on [28]the login page, but you can go there directly if you'd
   rather: [29]http://rgtp.thurman.org.uk/gossip/groggs/newbie. Once
   you've filled that in, you'll get email back explaining what to do
   next.
   
   If you find any problems, email the GROGGS editors; their address
   always appears in the [30]message of the day.
   
3. Getting started as an administrator

   (This section is a little more sketchy than I'd like. Any help would
   be appreciated.)
   
3.1. Installing yarrow

   There isn't (yet) any .deb or .rpm packaged version of yarrow, though
   it's on the drawing board. Instead, you'll need to [31]download the
   tarballs and install things yourself.
   
   You'll need a fairly recent version of [32]Python-- 2.1 should be good
   enough-- and (obviously) a working web server.
   
   You'll also need the Python Distutils installed. Many versions of
   Python have them installed already; I recommend, when you're following
   the instructions here, that you assume yours does until you find it
   doesn't. If yours turns out not to have them, you can find them in the
   Debian package [33]python2.1-dev (or [34]python2.2-dev, depending on
   the version of Python you're using; nothing's stopping you installing
   both). RPM users can [35]look on rpmfind. And everyone can get them
   from [36]the Python website.
   
   Unpack the tarball:
   
     tar xvzf yarrow-1.2.0.tar.gz
     
   and move into the directory it created:
   
     cd yarrow-1.2.0
     
   Now, as root, run
   
     make install
     
   If it all worked, then
   
     [37]http://localhost/cgi-bin/yarrow.cgi
     
   should be the address of a new and working RGTP server!
   
   Go ahead and test it by clicking on the entry for GROGGS in the menu,
   and then the "Yes, I'm just visiting" button. The page should take a
   good while to appear the first time, because yarrow has to download
   the whole index for the past seven years or so-- there's about half a
   megabyte. Fortunately, this only happens the first time you read from
   any server: after that, it's all cached.
   
   Some odds and ends:
     * If you don't have root access to the machine you're installing
       yarrow onto, you'll have to install things more or less yourself.
       You can tell yarrow where to find the files by modifying
       yarrow.conf; if you have to move yarrow.conf itself, simply change
       the definition of the variable "baseconf" in config.py.
     * You may also like to install some other RGTP clients; there are
       several listed on the [38]links page.
     * If you need any help, please do email me-- I need to know how this
       section can be improved.
     * Note that there is no gopher support in this version. I may add it
       back in if anyone asks for it.
       
3.2. Installing your own RGTP servers

   This is rather more difficult. There are, as far as I know, two RGTP
   server codebases in the world. One is the one [39]GROGGS runs on,
   which has never been officially released. The other is one I'm
   writing, called [40]spurge; it's been used effectively in the past,
   but it's in no way stable enough for a production environment. I hope
   that in the future it will be possible to use spurge to run a reliable
   RGTP server; in the meantime, you're welcome to download it and try it
   out. Offers of help and messages of encouragement are welcomed.
   
4. The future

   This project has largely met its original goals: I learned a good deal
   both about Python and about web applications, it was both challenging
   and fun to write, and it got a good number of people interested once
   more in GROGGS. It also sparked the imagination of others, bringing
   forth other projects in their own right (notably [41]Vendepacx). Now
   that all this is done, my hopes for its future are:
     * that it will serve as an accessible method of using GROGGS for a
       new generation of groggers
     * that it can be packaged into .deb and .rpm form, and, perhaps,
       become a package within the Debian project
     * that (combined with a stable server) it can become a viable option
       for people looking to run a small, fast and simple discussion
       system. In diversity is strength.
       
5. Credits

   [42]John Stark (jas11) created GROGGS. [43]Ian Jackson (iwj10) created
   RGTP. The interface of yarrow was largely modelled after
   [44]LiveJournal, which was created by [45]Brad Fitzpatrick. The
   following people, in alphabetical order, have given help, advice or
   moral support during the growth of yarrow: [46]Peter Colledge,
   [47]Natalie Currie, [48]Martin Hardcastle, [49]Michelle Hart, [50]Neil
   Roques, [51]Firinel Taranen, [52]Simon Tatham. Thanks to all of you.
   There are probably more people I've left out; if one of them's you,
   please do tell me.
   
yarrow

   [53]Downloads
   [54]Documentation
   [55]Screen shots
   [56]Links
   [57]Contact
   
Get more

   [58]Try it yourself
   [59]Freshmeat
   [60]CVS
   [61]Home
   
Site

   [62]valid HTML
   [63]valid CSS

References

   1. http://rgtp.thurman.org.uk/yarrow/docs/index.php#introduction
   2. http://rgtp.thurman.org.uk/yarrow/docs/index.php#user
   3. http://rgtp.thurman.org.uk/yarrow/docs/index.php#basics
   4. http://rgtp.thurman.org.uk/yarrow/docs/index.php#groggs
   5. http://rgtp.thurman.org.uk/yarrow/docs/index.php#glossary
   6. http://rgtp.thurman.org.uk/yarrow/docs/index.php#more
   7. http://rgtp.thurman.org.uk/yarrow/docs/index.php#admin
   8. http://rgtp.thurman.org.uk/yarrow/docs/index.php#clients
   9. http://rgtp.thurman.org.uk/yarrow/docs/index.php#servers
  10. http://rgtp.thurman.org.uk/yarrow/docs/index.php#future
  11. http://rgtp.thurman.org.uk/yarrow/docs/index.php#credits
  12. http://www.groggs.group.cam.ac.uk/groggs/
  13. http://rgtp.thurman.org.uk/yarrow/screenshots/
  14. http://rgtp.thurman.org.uk/yarrow/contact/
  15. http://rgtp.thurman.org.uk/yarrow/links/
  16. http://rgtp.thurman.org.uk/gossip
  17. http://www.groggs.group.cam.ac.uk/cgi-bin/yarrow.cgi
  18. http://rgtp.thurman.org.uk/gossip/groggs/browse
  19. http://rgtp.thurman.org.uk/gossip
  20. http://rgtp.thurman.org.uk/yarrow/screenshots/browse
  21. http://rgtp.thurman.org.uk/yarrow/screenshots/edit
  22. http://www.sandtreader.com/paul/groggs/springfield.html
  23. http://www.ucam.org/chat/
  24. http://www.chiark.greenend.org.uk/~owend/free/GREED.html
  25. http://www.chiark.greenend.org.uk/~sgtatham/putty/
  26. http://www.uni-oldenburg.de/fsphysik/dokumente/42/1.98/Hichhiker.html
  27. http://homepages.ed.ac.uk/cboyd/
  28. http://rgtp.thurman.org.uk/yarrow/screenshots/login
  29. http://rgtp.thurman.org.uk/gossip/groggs/newbie
  30. http://rgtp.thurman.org.uk/gossip/groggs/motd
  31. http://rgtp.thurman.org.uk/yarrow/download/
  32. http://www.python.org/

CHANGES
Changelog

v1.20
=====

1. If the browser supports it, the HTML that yarrow produces will be
compressed.

2. Fixed the config login bug: yarrow now no longer makes a connection
to the RGTP server when producing a configuration page. If, for example,
the server decided that the shared-secret was not the one yarrow knew
about, the connection attempt would fail. Since yarrow would then not
render the rest of the page, it was impossible to change the
shared-secret to the correct one.

3. A couple of enhancements for those running old or text-only browsers.
Unread articles in the index are prefixed with a degree sign, instead of
using the bold attribute to convey the information: some browsers don't
display this well, especially within a hyperlink. Replies to an item are
separated using a horizontal rule on non-CSS browsers: the solid bar of
the title line does this on browsers which support CSS.

4. The Content-Length header is set correctly, allowing the use of
persistent connections.

5. Introduced rudimentary caching support (ETags only, not yet using
Last-Modified).

6. The user can set an RGTP server's message of the day, if they have
the correct permissions.

7. The user can modify an RGTP server's user database, if they have the
correct permissions.

8. CSS is included in all HTML pages. Nowhere does yarrow now rely on static
content from the HTTP server.

9. Warnings that you need cookies added to the login pages. (Thanks to
cgb10 for this idea.)

10. Quite a bit of code tidying.

11. Installation script, based on distutils. No .rpm or .deb support yet.

v1.11
=====

1. Fixed a silly bug which stopped you being able to request a new account
on an RGTP server. A couple of variable names were being sent as literal
text.

2. Something that should have been in 1.10 but didn't make the cut: the
screen that tells you you must be logged in to read a server, and gives
you a button to do so, is sent with HTTP 200 OK, not HTTP 403 Forbidden.
This means that intermediate agents are less likely to stop the end user
seeing the content and so knowing how to fix it. It also means that such
a page can be, for example, set as the project's demonstration page on
Freshmeat.

v1.10
=====

1. A new "visitor" mode. Everyone who doesn't have a login cookie can get one
just by pressing a button, as well as by logging in. The visitor cookie isn't
good for much, but it gives you access to servers which allow guest login;
see the next point.

2. Automatic guest login has been reinstated if you're logged in at all, even
if it's only as a visitor (see previous point). This was removed by request
in v0.40, but I was far too draconian and removed it in _all_ cases, which
effectively stopped non-groggers from reading GROGGS at all. So you can now
browse as much as you like.

3. It's now much less important where you install yarrow. Only one path
is hard-wired (that of the configuration file) and that can be simply changed.
All other paths are given in the configuration.

4. Dates are now printed slightly more prettily: the "am/pm" is lowercase
and without a space. Also, rgtp IDs (email addresses) are printed in
monospace; I'm not sure I like this, but I'm experimenting with it.

5. URLs containing commas are properly linkified.

6. Fixed bug from v1.00: RGTP logs are printed only once. (They were being
printed twice, once after the closing </html> tag.)

7. The action of confirming the "catchup" option must be submitted by
HTTP PUT, so that it can't easily be faked by an external link.

v1.00
=====

Yep, we're finally here. Most of the changes were minor: v0.40 turned out to
have been the last big step, as the plan was.

Changes in this version:

1. Ability to page through the index, instead of being shown all the several
hundred entries on one page

2. Reformat option reinstated

3. Option to view the MOTD at the top of the index, or not, or only when it's
changed

4. Logging the RGTP session to screen reinstated

5. HTTP errors returned on error pages

6. Accepts server hostnames as well as nicknames in URLs (thanks to Peter
Colledge for the suggestion)

7. Manual written; website almost entirely rewritten too.

Happy grogging... 

v0.40
=====

This version has had more changes than any other version since the beginning.
Because of this, it's been a good long while since the release of 0.31, so I
thought a new release was about due once the code was fairly stable. Some
of the lesser-used features have been disabled or broken, though, and some of
the planned new features aren't yet working. I'm planning to move through
the v0.4x releases fairly quickly as I fix these problems.

This should, DV, be the last major release before v1.00.

Specific new features:

1. ALL STATE IS HELD ON THE SERVER. One password, which the user chooses, gets
them access to everything. In v0.31 and before, all state was held in the
browser, using cookies. This caused security problems (shared-secrets shouldn't
be sent across the net more than necessary, and certainly not on every pageview),
usability problems (the user must enter their shared-secret on every machine
they use, possibly more often than once if it's a public machine they can't
stay logged into) and feature problems (features that needed a lot of state,
such as (2) below, couldn't be implemented.)

2. yarrow remembers when the user last read an item, and prints the item's name
in bold if it's new or updated since then. (However, we may have to look for
another way of doing this: Lynx has trouble distinguishing bold and non-bold
hyperlinked text.)

3. Removed feature: earlier versions would attempt to log into servers as
"guest" if they couldn't get in any other way. This provides public access
to some servers (notably the Cambridge GROGGS system). However, people
were simultaneously concerned about privacy and the intrusion of robots,
and unwilling to remove the "guest" account for general access. So, in

rgtp.py
#!/usr/bin/python
#
#  rgtp.py - RGTP client library
#  v0.02
#
# Copyright (c) 2002 Thomas Thurman
# thomas@thurman.org.uk
# 
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have be able to view the GNU General Public License at 
# http://www.gnu.org/copyleft/gpl.html ; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307, USA.

import socket
import string
import md5
import binascii
import wrapping

###########################################################

class RGTPException (Exception):
	"Houston, we have a problem."

	def __init__(self, name):
		self.name = name

	def __str__(self):
		return self.name

###########################################################

class response:
	"One message from the server."

	def __init__(self, text, code=None):
		"Creates a response from a line of text received from the server."

		if code==None:
			try:
				self.numeric = int(text[0:3])
				self.textual = text[4:]
			except:
				self.numeric = -999
				self.textual = '(Weird bug finding:) ' +text
		else:
			self.numeric = code
			self.textual = text
		self.maybe_panic()

	def maybe_panic(self):
		if self.numeric==481:
			raise RGTPException("Timeout: "+self.textual)
		elif self.numeric==484 or self.numeric==-999:
			raise RGTPException("Server internal error: "+self.textual)
		elif self.numeric==500 or self.numeric==510 or self.numeric==511 or self.numeric==512 or self.numeric==582:
			raise RGTPException("Broken client: "+self.textual)
		elif self.numeric==484:
			raise RGTPException("Server internal error: "+self.textual)
		elif self.numeric==530 or self.numeric==531:
			raise RGTPException("Permission denied. (Try logging in with a privileged account?): "+self.textual)

	def code(self):
		return self.numeric

	def text(self):
		return self.textual

	def __str__(self):
		return str(self.numeric) + " " + self.textual

###########################################################

class callback:
	def __call__(self, message):
		pass

	def done(self):
		return 1

###########################################################

class expect(callback):
	def __init__(self, expectation):
		self.desideratum = expectation

	def __call__(self, message):
		if message.code() != self.desideratum:
			raise "Expected " + str(self.desideratum) + " but got " + message

###########################################################

class multiline(callback):

	def __init__(self):
		self.finished = 0

	def complete(self):
		self.finished = 1

	def done(self):
		return self.finished

###########################################################

class stomach(multiline):
	def __init__(self):
		multiline.__init__(self)
		self.stuff = []

	def __call__(self,message):
		self.eat(message)

	def eat(self,message):
		if message.code()==250:
			pass # data coming up-- good
		elif message.code()==-1:
			self.stuff.append(message.text())
		elif message.code()==0:
			self.complete()
		else:
			raise RGTPException("Wasn't expecting " + str(message))

###########################################################

class base:
	"Basic RGTP handling."

	def __init__(self, host, port, cback, logging):
		self.logging=logging
		self.log=''
		self.state = 0
		sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
		sock.connect((host, port))
		self.incoming = sock.makefile("r")
		self.outgoing = sock.makefile("w")
		self.cback = cback
		self.get_line()

	def get_line(self):
		looping = 1

		while looping:
			temp = ''

			# Read from the client until we get a response.
			while len(temp)==0:
				temp = self.receive()
				message = response(temp)
				self.cback(message)

			if message.code()==250: # Magic value for continuations
				while temp!='.':
					temp = self.receive()
					if temp!='.':
						self.cback(response(temp, -1))
				self.cback(response('', 0))

			# okay. Ask whether we should go round again.
			looping = not self.cback.done()

	def receive(self):
		temp = string.rstrip(self.incoming.readline())
		if self.logging: self.log = self.log + "\n<" +temp
		return temp

	def raw_send(self, message):
		"Simply sends one line to the server."
		self.outgoing.write(message + "\r\n")
		self.outgoing.flush()
		if self.logging:
			self.log = self.log + "\n>"+message

	def send(self, message, cback):
		"Sends one line to the server, and waits for a response."
		self.cback = cback
		self.raw_send(message)
		self.get_line()

###########################################################

class fancy:
	"Encapsulated RGTP."

	base = 0
	access_level = 0

	def __init__(self, host, port, logging):
		class first_connect(callback):
			access_level = 0
			def __call__(self, message):
				self.access_level = message.code()-230

		towel = first_connect()
		self.base = base(host, port, towel, logging)
		self.access_level = towel.access_level

	def login(self, email, sharedsecret = None):

		class authorise(multiline):

			def __init__(self, base, email, sharedsecret):
				multiline.__init__(self)
				self.clientnonce = "8a0eb22b27cc2dd5373f8cd9657fe8ea"
				self.hash = md5.new()
				self.base = base
				self.email = email[0:16]
				while len(self.email)<16:
					# pad with nuls if it's <16 bytes
					self.email += '\0'
				self.sharedsecret = sharedsecret

			def __call__(self, message):
				if message.code()==333:
					def inverted_bitstring(x):
						result = ""
						for i in range(len(x)):
							result = result + chr(255-ord(x[i]))
						return result
					self.hash.update(binascii.unhexlify(self.clientnonce))
					self.hash.update(binascii.unhexlify(message.text()))
					self.hash.update(self.email)
					self.hash.update(inverted_bitstring(binascii.unhexlify(self.sharedsecret)))
					self.base.send("AUTH "+self.hash.hexdigest()+" "+self.clientnonce, self)
				elif message.code()==133:
					pass # ummm...
				elif message.code()==483:
					raise RGTPException("Authentication failed ("+message.text()+")")
				elif message.code()==130:
					pass # ignore this
				elif message.code()>=230 and message.code()<=233:
					self.access_level = message.code()-230
					self.complete()
				elif message.code()==482 or message.code()==483 or message.code()==432 or message.code()==433:
					raise RGTPException("Failed to log you in - " + message.text())
				else:
					raise RGTPException("Wasn't expecting " + str(message))

		towel = authorise(self.base, email, sharedsecret)
		self.base.send("USER "+email, towel)
		self.access_level = towel.access_level

	def request_account(self, email):
		class regu_handler(stomach):

			def __init__(self):
				stomach.__init__(self)

			def __call__(self, message):
				if message.code()==100:
					pass # probably best to ignore this
				elif message.code()==482 or message.code()==280:
					successful = message.code()==280
					self.answer = (successful, message.text())
				elif message.code()==250:
					pass # good, that's what we want
				elif message.text()=='' or message.text()[0]==' ':
					self.eat(response(message.text()[1:], message.code()))

		towel = regu_handler()
		self.base.send("REGU", towel)
		if email!=None:
			self.base.send("USER "+email, towel)
			return towel.answer
		else:
			return towel.stuff

	def motd(self):
		towel = stomach()
		self.base.send("MOTD", towel)
		return towel.stuff

	def index(self):
		class index_reader(multiline):
			def __init__(self):
				self.result = []

			def __call__(self, message):
				if message.code()==-1:
					b = message.text()
					self.result.append((string.strip(b[0:8]), string.strip(b[9:17]),
						string.strip(b[18:26]), string.strip(b[27:102]),
						b[103], string.strip(b[105:])))
				elif message.code()==0:
					self.complete()

		towel = index_reader()
		self.base.send("INDX", towel)
		return towel.result

	def interpreted_index(self):
		index = {}
		for line in self.index():

			if line[4]=='I' or line[4]=='R' or line[4]=='C' or line[4]=='F':
				if not index.has_key(line[2]):
					index[line[2]] = {'date': 0, 'count': 0, 'subject': 'Unknown', 'live': 1 }

				if line[4]=='I' or line[4]=='C':
					index[line[2]]['subject'] = line[5]

				if index[line[2]]['date'] < int(line[1],16):
					index[line[2]]['date'] = int(line[1],16)
					index[line[2]]['from'] = line[3]

				if line[4]=='F':
					index[line[2]]['live'] = 0

				index[line[2]]['count'] = index[line[2]]['count'] + 1;

		return index


	def logout(self):
		if self.access_level != 0:
			self.base.send("QUIT", expect(280))
			self.access_level = 0

	def __del__(self):
		self.logout()

	def item(self, id):
		class item_reader(multiline):
			def __init__(self):
				multiline.__init__(self)
				self.result = []
				self.firstline = 0
				self.buffer = ''
				self.subject = ''

			def put_buffer(self):
				if self.buffer!='':
					lines = string.split(self.buffer, '\n')
					grogname = ''
					author = ''
					date = ''

					# Firstly, there'll be "item" or "reply" lines.
					# (Perhaps we should complain if there aren't.)
					if lines[0][0:5]=='Item ':
						date = lines[0][19:]
						lines = lines[1:]
					elif lines[0][0:11]=='Reply from ':
						date = lines[0][11:]
 						lines = lines[1:]

					atpos = string.rfind(date, ' at ')
					if atpos != -1:
						author = date[:atpos]
						date = date[atpos+4:]

					# "From" introduces an explicit grogname.
					if lines[0][0:5]=='From ':
						grogname = lines[0][5:]
						lines = lines[1:]

					# If the server tells us a subject, ignore it;
					# we'll have other ways of finding that out.
					if lines[0][0:9]=='Subject: ':
						lines = lines[1:]

					# Right. If we don't know the grogname by now,
					# it might have been short enough to go into the
					# author field.
					if grogname=='':
						openbracket = string.rfind(author, '(')
						if openbracket!=-1 and author[-1]==')':
							# Ah, so it was.
							grogname = author[:openbracket-1]
							author = author[openbracket+1:-1]
						else:
							# Well, just give them the address again.
							grogname = author

					self.result.append((grogname, date, author, lines))

					self.buffer=''

			def __call__(self, message):
				if message.code()==-1:
					text = message.text()
					if self.firstline:
						self.firstline = 0
						# should parse it, but...
						# FIXME: what's this about?
						self.result.append(text)
						self.result.append(text)
					elif text!='' and text[0]=='^' and text[1]!='^':
						self.put_buffer()
					else:
						if text!='' and text[0] in ['^', '.']:
							text = text[1:]
						self.buffer=self.buffer+text+"\n"
				elif message.code()==0:
					self.put_buffer()
					self.complete()
				elif message.code()==410:
					raise "No such item"
				elif message.code()==250:
					self.firstline = 1

		towel = item_reader()
		self.base.send("ITEM "+id, towel)
		return towel.result


        def raise_access_level(self, target=None, user=None, password=None):
		# Set target==None to get as high as we can with current
		# credentials.
                if target==None or target > self.access_level:
			# If they want more than they already have...
                        if user!=None and user!='':
				# They have a username. Fine: use it.
                                self.login(user, password)
		                if target!=None and target > self.access_level:
					raise RGTPException(user + " doesn't have a high enough access level.")
                        else:
				# No username. Hmm, maybe we can try the "guest" trick.
                                if (target==None or target==1) and self.access_level==0:
                                        self.login("guest", 0)
                                else:
					if target!=None:
	                                        raise RGTPException("You need to log in for that.")

		# So, did it work?
                if target > self.access_level:
			raise RGTPException("Sorry: try logging in with a more privileged account.");

	def stat(self, id):
		class status_reader(callback):
			def __call__(self, message):
				if message.code()==211:
					self.result = message.text()
				elif message.code()==410:
					raise RGTPException("Not available: "+message.text())
				else:
					raise RGTPException("Wasn't expecting " + str(message))

		def maybe_blank(thing):
			if thing=='        ':
				return None
			else:
				return thing

		towel = status_reader()
		self.base.send("STAT "+id, towel)
		r = towel.result
		return {'from': maybe_blank(r[0:8]), 'to': maybe_blank(r[9:17]), 'edited': maybe_blank(r[18:26]), 'replied': maybe_blank(r[27:35]), 'subject': r[36:] }

	def send_data(self, grogname, message):
		class dumper(multiline):

			def __init__(self, name, data, base):
                                multiline.__init__(self)
				self.name = name
				self.data = data
				self.base = base

			def __call__(self, message):
				def dot_doubled(line):
					if line!='' and line[0]=='.':
						# Dot-doubling: if it already
						# begins with a dot, it needs
						# another.
						return '.' + line
					else:
						return line

				if message.code()==150:
					# The server says "go ahead". So here goes!
					self.base.raw_send(dot_doubled(self.name))
					for paragraph in self.data:
						for line in wrapping.wrap(paragraph):
							self.base.raw_send(dot_doubled(line))
					# All done!
					self.base.raw_send('.')
				elif message.code()==350:
					self.complete()
				else:
					raise("Wasn't expecting " + str(message))

		self.base.send('DATA', dumper(grogname, message, self.base))

	def post(self, item, subject):
		# If item is None, this is a NEWI.
		# If item is not None and subject is None, this is a REPL.
		# If item is not None and subject is not None, this is a CONT.
		#
		# Returns:
		#  -1 -- failure because this item has been edited
		#   1 -- failure because this item is full now; you must CONT
		# a string -- success. the string is the itemid
		# (Is this ugly? Maybe we should return a tuple.)
		#
		# Obviously not all these can be returned in all cases.
		# You should check for the item having been edited already
		# yourself, as well, and not just rely on this function
		# returning -1; there are cases this function won't pick up
		# (such as editing which doesn't cause continuation). 

		class item_generator(multiline):
			def __init__(self, itemid):
                                multiline.__init__(self)
				self.itemid = itemid
				self.already_edited_status = 0

			def __call__(self, message):
				if message.code()==120:
					# A new itemid for us.
					self.itemid = message.text()
				elif message.code()==122:
					pass # for now. hmm.
				elif message.code()==220:
					# woohoo! all done
					self.complete()
				elif message.code()==421:
					self.already_edited_status = 1
					self.complete()
				elif message.code()==422:
					# it's overflowed!
					self.already_edited_status = -1
					self.complete()
				else:
					raise("Wasn't expecting " + str(message))

		towel = item_generator(item)
		if item==None:
			self.base.send('NEWI '+subject, towel)
		else:
			self.base.send('REPL '+item, towel)

			if towel.already_edited_status==1 and subject!=None:
				# we were given a subject in case this happened.
				# Use it.
				towel = item_generator(item)
				self.base.send('CONT '+subject, towel)

		if towel.already_edited_status == 0:
			return towel.itemid
		else:
			return towel.already_edited_status


